modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientComputeJobPacker.java

```diff
@@ -0,0 +1,144 @@
+package org.apache.ignite.internal.client.proto;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.Period;
+import java.util.UUID;
+import org.apache.ignite.internal.binarytuple.inlineschema.TupleWithSchemaMarshalling;
+import org.apache.ignite.marshalling.Marshaller;
+import org.apache.ignite.sql.ColumnType;
+import org.apache.ignite.table.Tuple;
+import org.jetbrains.annotations.Nullable;
+
+/** Packs job arguments and results. */
+public final class ClientComputeJobPacker {
+    /**
+     * Packs compute job argument. If the marshaller is provided, it will be used to marshal the argument. If the marshaller is not provided
+     * and the argument is a native column type or a tuple, it will be packed accordingly.
+     *
+     * @param arg Argument.
+     * @param marshaller Marshaller.
+     * @param packer Packer.
+     * @param <T> Argument type.
+     */
+    public static <T> void packJobArgument(@Nullable T arg, @Nullable Marshaller<T, byte[]> marshaller, ClientMessagePacker packer) {
+        pack(arg, marshaller, packer);
+    }
+
+    /**
+     * Packs compute job result. If the marshaller is provided, it will be used to marshal the result. If the marshaller is not provided and
+     * the result is a native column type or a tuple, it will be packed accordingly.
+     *
+     * @param res Result.
+     * @param marshaller Marshaller.
+     * @param packer Packer.
+     * @param <T> Result type.
+     */
+    public static <T> void packJobResult(@Nullable T res, @Nullable Marshaller<T, byte[]> marshaller, ClientMessagePacker packer) {
+        pack(res, marshaller, packer);
+    }
+
+    /** Packs object in the format: | typeId | value |. */
+    private static <T> void pack(@Nullable T obj, @Nullable Marshaller<T, byte[]> marshaller, ClientMessagePacker packer) {
+        if (marshaller != null) {
+            packer.packInt(ComputeJobType.MARSHALLED_OBJECT_ID);
+            byte[] marshalled = marshaller.marshal(obj);
+
+            if (marshalled == null) {
+                packer.packNil();
+                return;
+            }
+
+            packer.packByteBuffer(ByteBuffer.wrap(marshalled));
+            return;
+        }
+
+        if (obj instanceof Tuple) {
+            byte[] marshalledTuple = TupleWithSchemaMarshalling.marshal((Tuple) obj);
+
+            packer.packInt(ComputeJobType.MARSHALLED_TUPLE_ID);
+
+            if (marshalledTuple == null) {
+                packer.packNil();
+                return;
+            }
+
+            packer.packByteBuffer(ByteBuffer.wrap(marshalledTuple));
+            return;
+        }
+
+        if (obj == null) {
+            packer.packInt(ColumnType.NULL.id());
+            packer.packNil();
+            return;
+        }
+
+        var type = getColumnTypeFrom(obj);
+
+        packer.packInt(type.id());
```

Comment:
* This type id is not used during unpack
* It is duplicated as part of the binary tuple

I suggest to make `ComputeJobType.NATIVE` just another constant.